### PR TITLE
Remove persistence of RSS plans

### PR DESCRIPTION
**Pull request by andrewjstone**
Implements #7174

Obviates the need for #724

This still needs some manual testing.

### CODE REVIEW

File: sled-agent/src/rack_setup/service.rs [hunk 1]
```diff
@@ -291,6 +295,18 @@ impl RackSetupService {
...
+    fn is_newer_than(&self, _other: &Self) -> bool {
+        true
+    }
+    fn generation_bump(&mut self) {}
+}
+
+const RSS_STARTED_FILENAME: &str = "rss-started.marker";
+
 #[derive(Clone, Serialize, Deserialize, Default)]
 struct RssCompleteMarker {}
```
Comment (smklein) [inline sled-agent/src/rack_setup/service.rs:311]: Noting this for myself and others: The "RSS completed" marker is not changing with this PR, so systems which have fully deployed with RSS already will be unimpacted by this change. (This is good)

Comment (andrewjstone) [reply sled-agent/src/rack_setup/service.rs:311]: Yep, done on purpose. I thought about adding a comment in code but didn't end up doing it. Happy to if you think it will help though.

File: sled-agent/src/rack_setup/service.rs [hunk 2]
```diff
@@ -1450,13 +1419,21 @@ impl ServiceInner {
...
-            &plan,
+            &sled_plan,
             &service_plan,
             ExternalPortDiscovery::Auto(switch_mgmt_addrs),
             nexus_address,
         )
         .await?;
 
+        // Finally, mark that we've completed executing the plans and handed off to nexus.
+        let mut ledger = Ledger::<RssCompleteMarker>::new_with(
```
Comment (smklein) [inline sled-agent/src/rack_setup/service.rs:1430]: If we handoff to nexus, but crash before writing the ledger, wouldn't this system risk re-running RSS concurrently with Nexus changes (e.g., the reconfigurator)?

I know we can't make the "write ledger" and "handoff to nexus" steps atomic, but in the "old version" of this code, we did the following:

- Write down the `RssCompleteMarker` **first**
- Then handoff to Nexus
- If the RSS service crashes and reboots, it can handoff to Nexus again
- Nexus should know how to handle a handoff idempotently, even well after initialization, so this operation is safe and doesn't risk allocating services outside Nexus' control

Comment (smklein) [reply sled-agent/src/rack_setup/service.rs:1430]: If we don't want to do the handoff (e.g., maybe the serialization of this data isn't something we want to care about?) we could have a Nexus internal endpoint that's much simpler -- just asking "did handoff happen".

Then, this PR could:

- Write that RSS completed (`RssCompletedMarker`)
- Perform the handoff to Nexus
- Carry on

And in the case that we crash after writing `RssCompletedMarker`:

- RSS would read the marker on reboot
- RSS could query the "did handoff happen" endpoint (no parameters, just a boolean response)
- If yes: Cool, carry on
- If no: Throw the same error about needing a clean slate that we would throw if RSS had started, but not completed

Comment (smklein) [reply sled-agent/src/rack_setup/service.rs:1430]: (this could also help with the bootstore proposal you wrote up - we wouldn't need to store the entire RSS config, just a boolean indicating that we **should** be done with RSS)

Comment (andrewjstone) [reply sled-agent/src/rack_setup/service.rs:1430]: Great feedback @smklein. Indeed, I was somewhat conflicted here. I like the existing logic of being able to idempotently handoff to nexus if the `RssCompleteMarker` was written already. But that also did indeed require persisting the plans, which adds more opportunity for things to go wrong and cognitive overhead for developers. I think we could do away with the versioning entirely and just persist the latest version of the plans and that would achieve the primary goal of #7174. I still would prefer not to persist the plans though, as you have surmised.

I like the design of your solution, but it suffers from the problem of having to indefinitely wait for an answer from Nexus or timeout and return an error. I'm also not sure the problem the new nexus API solves is really a problem though. I added the write of a start marker for this case and so we'd just clean slate if this happens. Basically, all the plan persistence and early completed marker write prevented was handling a crash after plan execution but before nexus handoff. 

This PR says "if we ever crash in RSS, we can't restart". We may have handed off, we may not have. In either case we accept what we have and clean slate. The gist of this PR is essentially a value judgement that restarting in case of this crash is cheaper and less problematic than persisting plan files. That is certainly open to interpretation.

Now, after saying that, I think there is another alternative, utilizing the bootstore. If nexus handoff actually did complete, but the marker didn't get written (if we keep the order as in this PR), Nexus could write to the bootstore to mark it complete. This would allow us to handle this crash in a similar manner to what we have now.

Comment (andrewjstone) [reply sled-agent/src/rack_setup/service.rs:1430]: > This would allow us to handle this crash in a similar manner to what we have now.

This isn't really true. But it does at least resolve the handoff completed, but marker not written case - given time for propagation.

Comment (smklein) [reply sled-agent/src/rack_setup/service.rs:1430]: > it suffers from the problem of having to indefinitely wait for an answer from Nexus or timeout and return an error.

Yeah, I don't view this as a problem for the case where RSS is initially doing handoff (it **should** be blocked on talking to Nexus!) but agreed that it could be a pain for the case where we just want to check "has RSS already run", which happens on cold boot.

As long as we say "the process of clean slating may wrench away control from a fully running Nexus, and we're okay with that", I agree that this new format is fine.

> Now, after saying that, I think there is another alternative, utilizing the bootstore. If nexus handoff actually did complete, but the marker didn't get written (if we keep the order as in this PR), Nexus could write to the bootstore to mark it complete. This would allow us to handle this crash in a similar manner to what we have now.

Yeah, and that's also nice because it propagates to the other scrimlet, right? Having the bootstore is handy, and seems like a nice addition in this part of the stack

Comment (andrewjstone) [reply sled-agent/src/rack_setup/service.rs:1430]: > > it suffers from the problem of having to indefinitely wait for an answer from Nexus or timeout and return an error.
> 
> Yeah, I don't view this as a problem for the case where RSS is initially doing handoff (it **should** be blocked on talking to Nexus!) but agreed that it could be a pain for the case where we just want to check "has RSS already run", which happens on cold boot.
> 

Agree. 

> As long as we say "the process of clean slating may wrench away control from a fully running Nexus, and we're okay with that", I agree that this new format is fine.
> 

We already have to park the rack before clean-slating, which stops all the nexuses, so this shouldn't introduce any new behavior here. 

> > Now, after saying that, I think there is another alternative, utilizing the bootstore. If nexus handoff actually did complete, but the marker didn't get written (if we keep the order as in this PR), Nexus could write to the bootstore to mark it complete. This would allow us to handle this crash in a similar manner to what we have now.
> 
> Yeah, and that's also nice because it propagates to the other scrimlet, right? Having the bootstore is handy, and seems like a nice addition in this part of the stack

Agreed. I will be back in the bootstore /trust quorum area very shortly. Interestingly, once we have full trust quorum, the gossip can actually be deterministic per key. This is because with full trust quorum the entire membership is known at each node. We can therefore store a bitmap (just a u32 - yay for 32 sleds) that indicates which nodes have seen the latest version. We refresh the bitmaps and gossip again when membership changes.

Comment (jgallagher) [reply sled-agent/src/rack_setup/service.rs:1430]: > We already have to park the rack before clean-slating, which stops all the nexuses, so this shouldn't introduce any new behavior here.

Minor quibble: I don't think we explicitly park the rack before clean-slating; it's more than clean-slating has a side effect of parking the rack (as removing all the services also inherently stops them from running). But I think Sean's phrasing of "the process of clean slating may wrench away control from a fully running Nexus, and we're okay with that" is certainly correct: the point of clean-slating is to wrench control away from _everything_.

I'm not sure how this might change with a wicket-exposed clean-slate. But all of this seems fine with the manual clean slate we have today.

File: sled-agent/src/rack_setup/service.rs [hunk 3]
```diff
@@ -16,31 +16,30 @@
...
+//! us pick up and restart at certain given points we instead only record two
+//! marker files. These marker files indicate one of two things:
 //!
-//! The service plan file has "-v2" in the filename because its structure
-//! changed in omicron#4466.  It is possible that on startup, RSS finds an
-//! older-form service plan.  In that case, it fails altogether.  We do not
-//! expect this condition to happen in practice.  See the implementation for
-//! details.
+//!   * RSS has run already but failed to complete. We must clean-slate the
+//!     rack if we try to run RSS again and see this file.
```
Comment (jgallagher) [inline sled-agent/src/rack_setup/service.rs:28]: The marker also exists on _successful_ RSS runs, right (i.e., nothing removes it)? Maybe reword this to something like

```suggestion
//!   * RSS has started. We must clean-slate the rack if we try to start
//!      RSS and see this file, as it indicates the original RSS attempt failed.
```

File: sled-agent/src/rack_setup/service.rs [hunk 4]
```diff
@@ -1192,122 +1208,93 @@ impl ServiceInner {
...
+        .await;
+        let completed_ledger = Ledger::<RssCompleteMarker>::new(
+            &self.log,
+            completed_marker_paths.clone(),
+        )
+        .await;
 
         // Check if a previous RSS plan has completed successfully.
         //
         // If it has, the system should be up-and-running.
```
Comment (jgallagher) [inline sled-agent/src/rack_setup/service.rs:1240]: Maybe expand or reword this comment slightly to cover the new `else` block?

**Review by jgallagher — APPROVED** [review]

File: sled-agent/src/rack_setup/service.rs (hunk 4 above)
Comment (andrewjstone) [reply sled-agent/src/rack_setup/service.rs]: Good call. Done in [ddb7cc3](https://github.com/oxidecomputer/omicron/pull/7185/commits/ddb7cc324ca822f24e35587051cc4e78b95c7b47)

File: sled-agent/src/rack_setup/service.rs (hunk 3 above)
Comment (andrewjstone) [reply sled-agent/src/rack_setup/service.rs]: Good call. Done in [ddb7cc3](https://github.com/oxidecomputer/omicron/pull/7185/commits/ddb7cc324ca822f24e35587051cc4e78b95c7b47)

**Comment by andrewjstone** [comment]
Testing: Rack init succeeded on madrid and the console is available.
